eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs

```diff
@@ -39,251 +41,235 @@ internal HttpMessage CreateSayHiRequest(string headParameter, string queryParame
                 uri.AppendQuery("optionalQuery", optionalQuery, true);
             }
             request.Uri = uri;
-            request.Headers.Add("head-parameter", headParameter);
-            request.Headers.Add("Accept", "application/json");
+            request.Headers.SetValue("head-parameter", headParameter);
```

Comment:
It's probably best to keep it as `Set` in unbranded -- that is a good point about perf in Azure clients being why we do `Add` there -- thanks for tracking that down @JoshLove-msft!  

I think for these methods, it shouldn't make a difference one way or another (apart from perf considerations), because the request is created in the method so there is no opportunity for end-users/others to provide headers that get overwritten.  `Set` is a little more robust to possible errors in generated code as the last writer will win and we wouldn't end up with headers being duplicated.

For now I think it's probably preferrable to have Azure clients use `Add` and unbranded use `Set` for reasons above -- we can look at whether we can improve the perf of `Set` in Azure.Core to be able to better align the two.

Reply:
> Set is a little more robust to possible errors in generated code as the last writer will win and we wouldn't end up with headers being duplicated on the request.

If we think Set is more robust to errors, then doesn't that imply there could be cases where there are multiple headers values specified for the same header? In which case, wouldn't we want to include all the values rather than having the last one win?

Reply:
> In which case, wouldn't we want to include all the values rather than having the last one win?

If that can be specified in TypeSpec, then I think it is fair to say we would want to all values for a given header to be included on a request before we add end-user headers to it.  And I can see why it might be nice from a generator perspective to delegate the header value concatenation logic to a Core type, rather than duplicating the logic in the generator as well.  I'm not sure I agree, though, that it make sense for header value concatenation to happen at runtime in the client, each time a user calls a service method that creates a request with multiple values for a given header specified in the TypeSpec.  In unbranded clients, we first call `Set` for header values that come from the service description, and then later either set or add headers an end-user has provided on the RequestOptions type (via the call to `message.Apply(options)`).  At some point we will update RequestContext to provide the same functionality in Azure clients.

Reply:
I don't think we would want to duplicate the concatenation logic in the generator, but I'm not sure if this is even a real scenario - it looks like the Typespec compiler complains if you attempt to use the same header name twice in an operation. Sounds like the recommendation is to use Set generally, but given the perf limitations with the Azure implementation of Set, we should revert to using Add for the Azure plugin, @live1206. Is there an issue tracking the perf of using Set in Azure.Core? If not, I'm happy to file it. Thanks @annelo-msft!

Reply:
> Is there an issue tracking the perf of using Set in Azure.Core? If not, I'm happy to file it. 

Have you measured the perf to identify the issue you are describing?

I am wondering because, looking at the code, you are correct that the abstract base class `Request` implements `SetHeader` by calling `RemoveHeader` and `AddHeader`.  I expect this was either to reduce the number of abstract methods derived types would be required to implement, or because `SetHeader` was added after the first GA of `Request` and it would have been a breaking change to add an abstract method to the base type.  `SetHeader` is virtual, though, and the implementation in the most common paths overrides it in what looks -- without measuring -- like a way that is equivalent or more efficient than the call to `AddHeader`.  So, unless users provide a custom transport, I would actually expect the call to `SetHeader` to have either equivalent or faster perf in most cases.

So maybe then we want to have Azure clients call `SetHeader` to align with unbranded like @live1206 is doing here?

Reply:
Thanks for the discussion!
As @annelo-msft mentioned, the concrete implementation override looks fine: [HttpWebRequestTransportRequest.SetHeader](https://github.com/Azure/azure-sdk-for-net/blob/ea5cb401d0caace6fab6bcbde610baaf79c26e07/sdk/core/Azure.Core/src/Pipeline/HttpWebRequestTransport.Request.cs#L30) and [MemoryRequest.SetHeader](https://github.com/Azure/azure-sdk-for-net/blob/ea5cb401d0caace6fab6bcbde610baaf79c26e07/sdk/core/Azure.Core/src/Shared/Multipart/MemoryRequest.cs#L32)
I will revert this change to use `Add` instead of `Setheader` to follow the current behavior of [Azure client](https://github.com/Azure/azure-sdk-for-net/blob/ea5cb401d0caace6fab6bcbde610baaf79c26e07/sdk/compute/Azure.ResourceManager.Compute/src/Generated/RestOperations/CapacityReservationsRestOperations.cs#L72).

Reply:
After the above discussion, I believe the correct thing is to align with the unbranded generator and use `SetHeader` as you are currently doing in this PR.  Please work with @JoshLove-msft to verify that that is the guidance from the points of contact on the generator team.

Reply:
Updated to `SetValue`